sdk/core/src/main/java/com/klaviyo/coresdk/networking/KlaviyoApiClient.kt

```diff
@@ -0,0 +1,180 @@
+import com.klaviyo.coresdk.networking.requests.KlaviyoApiRequest
+import com.klaviyo.coresdk.networking.requests.TrackApiRequest
+import java.util.concurrent.ConcurrentLinkedQueue
+import org.json.JSONArray
+import org.json.JSONObject
+
+/**
+ * Coordinator of API request traffic
+ */
+internal object KlaviyoApiClient : ApiClient {
+    internal const val QUEUE_KEY = "klaviyo_api_request_queue"
+
+    private val handlerThread = HandlerUtil.getHandlerThread(KlaviyoApiClient::class.simpleName)
+    private var handler: Handler? = null
+    private var apiQueue = ConcurrentLinkedQueue<KlaviyoApiRequest>()
+
+    init {
+        startListeners()
+        restoreQueue()
+    }
+
+    fun startListeners() {
+        // Flush queue immediately when app stops
+        Registry.lifecycleMonitor.onActivityEvent {
+            if (it is ActivityEvent.AllStopped) flushQueue()
+        }
+
+        // Flush queue when network connection is restored
+        Registry.networkMonitor.onNetworkChange { isOnline ->
+            if (isOnline) flushQueue()
+        }
+    }
+
+    override fun enqueueProfile(profile: Profile) {
+        enqueueRequest(IdentifyApiRequest(profile))
+    }
+
+    override fun enqueueEvent(event: Event, profile: Profile) {
+        enqueueRequest(TrackApiRequest(event, profile))
+    }
+
+    /**
+     * Enqueues an [KlaviyoApiRequest] to run in the background
+     * These requests are sent to the Klaviyo asynchronous APIs
+     *
+     * This method will initialize the API queue and the batching thread
+     * if this is the first request made since launch.
+     */
+    fun enqueueRequest(vararg requests: KlaviyoApiRequest) {
+        if (apiQueue.isEmpty()) {
+            initBatch()
+        }
+
+        for (request in requests) {
+            if (!apiQueue.contains(request)) {
+                apiQueue.offer(request)
+            }
+            Registry.dataStore.store(request.uuid, request.toJson())
+        }
+
+        Registry.dataStore.store(
+            QUEUE_KEY,
+            JSONArray(apiQueue.map { it.uuid }).toString()
+        )
+    }
+
+    /**
+     * Awaken queued requests from persistent store
+     */
+    fun restoreQueue() {
+        // TODO JSON type safety checks
+        Registry.dataStore.fetch(QUEUE_KEY)?.let {
+            val queue = JSONArray(it)
+            Array(queue.length()) { i -> queue.optString(i) }
+        }?.map { uuid ->
+            Registry.dataStore.fetch(uuid)?.let { json ->
+                val jsonObject = JSONObject(json)
+                val request = KlaviyoApiRequest.fromJson(jsonObject)
+                enqueueRequest(request)
+            }
+        }
+    }
+
+    /**
+     * Initialize the network queue batching with a thread to run on
+     */
+    private fun initBatch() {
+        if (!handlerThread.isAlive) {
+            handlerThread.start()
+            handler = HandlerUtil.getHandler(handlerThread.looper)
```

Comment:
hmm...surprised a little by this. I thought maybe there were newer concurrency mechanisms in kotlin.

Reply:
Also does this support cancellation?

Reply:
There is, coroutines, it could be on our later roadmap to update to use coroutines, but I didn't think this approach was outdated or problematic so I kept it as is (this was in the 2020 code)

Reply:
Pretty sure it supports cancellation yes, this is the main reason i reopened the network layer ticket though, I hadn't fully finished making this ApiClient class respond to lifecycle events properly.

Reply:
Does support cancelling "messages" see #33